src/views/projects/SourceBrowser/FileDiff.svelte

```diff
@@ -193,19 +377,40 @@
       {#if file.diff.type === "plain"}
         {#if file.diff.hunks.length > 0}
           <table class="diff">
-            {#each file.diff.hunks as hunk}
-              <tr class="diff-line">
+            {#each file.diff.hunks as hunk, hunkIdx}
+              <tr class="diff-line" data-type={headerSign(selection, hunkIdx)}>
                 <td colspan={2} />
                 <td colspan={6} class="diff-expand-header">
                   {hunk.header}
                 </td>
               </tr>
-              {#each hunk.lines as line}
-                <tr class="diff-line" data-expanded data-type={lineSign(line)}>
-                  <td class="diff-line-number left" data-type={lineSign(line)}>
+              {#each hunk.lines as line, lineIdx}
+                <tr
+                  class="diff-line"
+                  data-expanded
+                  data-type={lineSign(
+                    line,
+                    isLineSelected(selection, hunkIdx, lineIdx),
+                  )}>
+                  <td
+                    id={file.path + "::H" + hunkIdx + "L" + lineIdx}
+                    class="diff-line-number left"
+                    on:click={e => {
+                      updateHash(
+                        calcNewHash(file.path, hunkIdx, lineIdx, e.shiftKey),
+                      );
+                    }}
+                    data-type={lineSign(line)}>
                     {lineNumberL(line)}
                   </td>
-                  <td class="diff-line-number right" data-type={lineSign(line)}>
+                  <td
+                    class="diff-line-number right"
+                    on:click={e => {
+                      updateHash(
+                        calcNewHash(file.path, hunkIdx, lineIdx, e.shiftKey),
+                      );
```

Comment:
You can merge these to function into one. I’d also suggest passing the whole event instead of `e.shiftKey` and have the function choose how to handle modifier keys.

Reply:
can do this, worth the squeeze?
- need it to be two functions cause of `return`s, otherwise would be unreadable `else if`s
- also notice `e: any`

```diff
diff --git a/src/views/projects/SourceBrowser/FileDiff.svelte b/src/views/projects/SourceBrowser/FileDiff.svelte
index a133f29..4617dd9 100644
--- a/src/views/projects/SourceBrowser/FileDiff.svelte
+++ b/src/views/projects/SourceBrowser/FileDiff.svelte
@@ -22,7 +22,7 @@
   onMount(() => {
     window.addEventListener("keydown", e => {
       if (e.key === "Escape") {
-        updateHash("");
+        _updateHash("");
       }
     });
     window.addEventListener("hashchange", updateSelection);
@@ -152,36 +152,39 @@
     );
   }

-  function calcNewHash(
-    path: string,
-    hunkIdx: number,
-    lineIdx: number,
-    shift: boolean,
-  ): string {
+  function updateHash(path: string, hunkIdx: number, lineIdx: number, e: any) {
     // single line selection
-    if (!shift) {
-      return path + ":H" + hunkIdx + "L" + lineIdx;
+    if (!e.shiftKey) {
+      return _updateHash(path + ":H" + hunkIdx + "L" + lineIdx);
     }

     if (!selection) {
-      return "";
+      return _updateHash("");
     }

     // range selection
     if (hunkIdx === selection.startHunk) {
       if (lineIdx >= selection.startLine) {
-        return `${path}:H${hunkIdx}L${selection.startLine}H${hunkIdx}L${lineIdx}`;
+        return _updateHash(
+          `${path}:H${hunkIdx}L${selection.startLine}H${hunkIdx}L${lineIdx}`,
+        );
       } else {
-        return `${path}:H${hunkIdx}L${lineIdx}H${hunkIdx}L${selection.startLine}`;
+        return _updateHash(
+          `${path}:H${hunkIdx}L${lineIdx}H${hunkIdx}L${selection.startLine}`,
+        );
       }
     } else if (hunkIdx < selection.startHunk) {
-      return `${path}:H${hunkIdx}L${lineIdx}H${selection.startHunk}L${selection.startLine}`;
+      return _updateHash(
+        `${path}:H${hunkIdx}L${lineIdx}H${selection.startHunk}L${selection.startLine}`,
+      );
     } else {
-      return `${path}:H${selection.startHunk}L${selection.startLine}H${hunkIdx}L${lineIdx}`;
+      return _updateHash(
+        `${path}:H${selection.startHunk}L${selection.startLine}H${hunkIdx}L${lineIdx}`,
+      );
     }
   }

-  function updateHash(newHash: string) {
+  function _updateHash(newHash: string) {
     if (newHash !== "") {
       window.location.hash = newHash;
     } else {
@@ -396,9 +399,7 @@
                     id={file.path + "::H" + hunkIdx + "L" + lineIdx}
                     class="diff-line-number left"
                     on:click={e => {
-                      updateHash(
-                        calcNewHash(file.path, hunkIdx, lineIdx, e.shiftKey),
-                      );
+                      updateHash(file.path, hunkIdx, lineIdx, e);
                     }}
                     data-type={lineSign(line)}>
                     {lineNumberL(line)}
@@ -406,9 +407,7 @@
                   <td
                     class="diff-line-number right"
                     on:click={e => {
-                      updateHash(
-                        calcNewHash(file.path, hunkIdx, lineIdx, e.shiftKey),
-                      );
+                      updateHash(file.path, hunkIdx, lineIdx, e);
                     }}
                     data-type={lineSign(line)}>
                     {lineNumberR(line)}
```

Reply:
> * need it to be two functions cause of `return`s, otherwise would be unreadable `else if`s

Yes, you’re right. But no need to overthink this. How about this?

```diff
diff --git a/src/views/projects/SourceBrowser/FileDiff.svelte b/src/views/projects/SourceBrowser/FileDiff.svelte
index eba16a8..24f8daa 100644
--- a/src/views/projects/SourceBrowser/FileDiff.svelte
+++ b/src/views/projects/SourceBrowser/FileDiff.svelte
@@ -144,14 +144,14 @@
     }
   }
 
-  function calcNewHash(
-    path: string,
+  function hashFromSelection(
     hunkIdx: number,
     lineIdx: number,
-    shift: boolean,
+    event: MouseEvent,
   ): string {
+    const path = file.path;
     // single line selection
-    if (!shift) {
+    if (!event.shiftKey) {
       return path + ":H" + hunkIdx + "L" + lineIdx;
     }
 
@@ -173,6 +173,10 @@
     }
   }
 
+  function selectLine(hunkIdx: number, lineIdx: number, event: MouseEvent) {
+    updateHash(hashFromSelection(hunkIdx, lineIdx, event));
+  }
+
   function updateHash(newHash: string) {
     if (newHash !== "") {
       window.location.hash = newHash;
@@ -392,21 +396,13 @@
                   <td
                     id={[file.path, "H" + hunkIdx, "L" + lineIdx].join("-")}
                     class="diff-line-number left"
-                    on:click={e => {
-                      updateHash(
-                        calcNewHash(file.path, hunkIdx, lineIdx, e.shiftKey),
-                      );
-                    }}
+                    on:click={e => selectLine(hunkIdx, lineIdx, e)}
                     data-line-type={lineSign(line)}>
                     {lineNumberL(line)}
                   </td>
                   <td
                     class="diff-line-number right"
-                    on:click={e => {
-                      updateHash(
-                        calcNewHash(file.path, hunkIdx, lineIdx, e.shiftKey),
-                      );
-                    }}
+                    on:click={e => selectLine(hunkIdx, lineIdx, e)}
                     data-line-type={lineSign(line)}>
                     {lineNumberR(line)}
                   </td>
```

Not a huge deal though. Feel free to leave it like it is.

Reply:
https://github.com/radicle-dev/radicle-interface/pull/942/commits/55a71d1c651b38c6715fd67b97e23b5da97d6d6a